### PR TITLE
add title to speedreader documents

### PR DESCRIPTION
**Pull request by pes10k**
This is a very hacky, but i think correct, way of adding a title element to speedreader documents.  It does so through JS injection, which sidesteps all the tricky speedreader bits 

fixes brave/brave-browser#19182

### CODE REVIEW

**Comment by pes10k** [comment]
Note, i get a crash on nightly whenever i interact with the speedreader dialog in this build, but… i get that w/ and w/o my change, so i think (🤞 ) its unrelated to this change

**Review by antonok-edm — COMMENTED** [review]
I think [here](https://github.com/brave/brave-core/blob/master/components/speedreader/rust/lib/src/readability/src/extractor.rs#L250-L256) might be a better place to add the title element :stuck_out_tongue_closed_eyes: 

Something like this perhaps?

```rust
    if !meta.title.is_empty() {
        let title_blob = format!("<title>{}</title>", &meta.title);
        content = title_blob + &content;
    }
```

**Review by rillian — CHANGES_REQUESTED** [review]
Thanks for writing a patch!

I agree with Anton, doing this on the rust side is cleaner, and will have a smaller effect on page load.

**Comment by pes10k** [comment]
@rillian @antonok-edm okie dokie, ive reworked the patch to use @antonok-edm's much-better suggestion.  Review appreciated!

**Review by rillian — APPROVED** [review]
LGTM.

I'm sceptical this comes out correctly for non-utf8 titles, but still an improvement!

**Comment by pes10k** [comment]
okay, i think ready for review @iefremov (i added a CSP policy to prevent any script from running, and restricted the CSS that could be run, after talking with the security team)

File: components/speedreader/speedreader_rewriter_service.cc [hunk 1]
```diff
@@ -36,7 +40,13 @@ std::string GetDistilledPageStylesheet(const base::FilePath& stylesheet_path) {
...
   }
 
-  return "<style id=\"brave_speedreader_style\">" + stylesheet + "</style>";
+  const std::string style_hash = crypto::SHA256HashString(stylesheet);
+  const std::string style_hash_b64 = base::Base64Encode(base::as_bytes(
+    base::make_span(style_hash)));
+
+  return "<meta http-equiv=\"Content-Security-Policy\" content=\""
+            "script-src 'none'; style-src 'sha256-" + style_hash_b64 + "'\">\n"
+         "<style id=\"brave_speedreader_style\">" + stylesheet + "</style>";
```
Comment (rillian) [inline components/speedreader/speedreader_rewriter_service.cc]: This won't apply to the initial load of the built-in stylesheet in the `SpeedreaderRewriterService` constructor. Probably this code should move to a `WrapStylesheet` helper which is called both here and when `content_stylesheet_` is set directly in the constructor.

Comment (pes10k) [reply components/speedreader/speedreader_rewriter_service.cc]: okay i can change accordingly, though i've yet to find a case where the `content_stylesheet_` property is used ;)

Comment (pes10k) [reply components/speedreader/speedreader_rewriter_service.cc]: done and done https://github.com/brave/brave-core/pull/11545/commits/3c5fb368e1e3606015bee801e019a9727367aa1d

**Review by rillian — APPROVED** [review]

File: components/speedreader/speedreader_rewriter_service.cc [hunk 2]
```diff
@@ -19,13 +21,27 @@
 #include "brave/components/speedreader/speedreader_component.h"
 #include "brave/components/speedreader/speedreader_util.h"
 #include "components/grit/brave_components_resources.h"
+#include "crypto/sha2.h"
```
Comment (iefremov) [inline components/speedreader/speedreader_rewriter_service.cc:24]: please add `crypto` to BUILD.gn and DEPS

**Comment by pes10k** [comment]
okie dokie @iefremov , i think this should be good to go now

**Review by iefremov — APPROVED** [review]

**Comment by iefremov** [comment]
Thanks @pes10k for fixing!

**Pull request merged**